### PR TITLE
Change AF_UNSPEC to "AF_INET | AF_INET6"

### PR DESCRIPTION
**Pull request by xiong-gang**
Sometimes query gets hung on Redhat 7, the stack is like:
...

netstat shows it's related to NETLINK.

### CODE REVIEW

**Comment by yydzero** [comment]
LGTM

File: src/backend/cdb/motion/ic_common.c [hunk 1]
```diff
@@ -1041,7 +1041,7 @@ SendDummyPacket(void)
 
 	MemSet(&hint, 0, sizeof(hint));
 	hint.ai_socktype = SOCK_DGRAM;
-	hint.ai_family = AF_UNSPEC; /* Allow for IPv4 or IPv6  */
+	hint.ai_family = AF_INET | AF_INET6; /* Allow for IPv4 or IPv6  */
```
Comment (hlinnaka) [inline src/backend/cdb/motion/ic_common.c]: You cannot specify multiple families by ORing them together.

Comment (xiong-gang) [inline src/backend/cdb/motion/ic_common.c]: You are right. Sorry.

**Comment by xiong-gang** [comment]
Seems like the latest glibc fixed the bug in v2.23 already.
Close this PR.


"* getaddrinfo now detects certain invalid responses on an internal netlink
  socket.  If such responses are received, an affected process will
  terminate with an error message of "Unexpected error <number> on netlink
  descriptor <number>" or "Unexpected netlink response of size <number> on
  descriptor <number>".  The most likely cause for these errors is a
  multi-threaded application which erroneously closes and reuses the netlink
  file descriptor while it is used by getaddrinfo."

**Comment by hlinnaka** [comment]
On 02/03/16 07:13, xiong-gang wrote:
> Seems like the latest glibc fixed the bug in v2.23 already.
> Close this PR.
>
>
> "* getaddrinfo now detects certain invalid responses on an internal netlink
>    socket.  If such responses are received, an affected process will
>    terminate with an error message of "Unexpected error <number> on netlink
>    descriptor <number>" or "Unexpected netlink response of size <number> on
>    descriptor <number>".  The most likely cause for these errors is a
>    multi-threaded application which erroneously closes and reuses the netlink
>    file descriptor while it is used by getaddrinfo."

1. That seems unrelated to the getaddrinfo() call we're making. We don't 
actually mean to create a netlink socket, do we?

2. Per the above description, glibc will now *terminate the process* if 
it detects such an invalid response. That means PANIC in Postgres/GPDB. 
Surely that's even worse than a hung query.

Have you been able to reproduce the original problem?

- Heikki

**Comment by xiong-gang** [comment]
Hi Heikki,

Check this out:
https://sourceware.org/bugzilla/show_bug.cgi?id=12926
The call stack is the same as what i found in my test environment.
We don't mean to create a netlink socket, but as we specifying AF_UNSPEC, it ends up hanging there doing some netlink stuff.

I can reproduce it sometimes, it stop showing up since i change it to "AF_INET | AF_INET6" (which is wrong).

**Comment by hlinnaka** [comment]
On 02/03/16 07:56, xiong-gang wrote:
> Hi Heikki,
>
> Check this out:
> https://sourceware.org/bugzilla/show_bug.cgi?id=12926
> The call stack is the same as what i found in my test environment.
> We don't mean to create a netlink socket, but as we specifying AF_UNSPEC, it ends up hanging there doing some netlink stuff.
>
> I can reproduce it sometimes, it stop showing up since i change it to "AF_INET | AF_INET6" (which is wrong).

And can you still reproduce it if you upgrade glibc to v2.23?

- Heikki

**Comment by xiong-gang** [comment]
update the glibc patch and the problem is still there.

@hlinnaka here's some information about the problem, do you have any idea about it?

call stack:
#0  0x00007f2cadb881fd in recvmsg () from /lib64/libc.so.6
#1  0x00007f2cadbaaa6d in make_request () from /lib64/libc.so.6
#2  0x00007f2cadbaaf64 in __check_pf () from /lib64/libc.so.6
#3  0x00007f2cadb71029 in getaddrinfo () from /lib64/libc.so.6
#4  0x00000000009fb949 in setupUDPListeningSocket (listenerSocketFd=listenerSocketFd@entry=0x18e839c, listenerPort=listenerPort@entry=0x7fffc3da7070, txFamily=txFamily@entry=0x18e83a0)
    at ic_udpifc.c:1232
#5  0x0000000000a086f2 in startOutgoingUDPConnections (pOutgoingCount=<synthetic pointer>, sendSlice=<optimized out>, transportStates=<optimized out>) at ic_udpifc.c:2989
#6  SetupUDPIFCInterconnect_Internal (estate=<optimized out>) at ic_udpifc.c:3462
#7  SetupUDPIFCInterconnect (estate=<optimized out>) at ic_udpifc.c:3523
#8  0x00000000009e52a3 in SetupInterconnect (estate=estate@entry=0x1980188) at ic_common.c:741
#9  0x0000000000678bf7 in ExecutorStart (queryDesc=queryDesc@entry=0x1936ba8, eflags=0) at execMain.c:516
#10 0x000000000085296f in PortalStart (portal=portal@entry=0x18d5ac8, params=params@entry=0x0, snapshot=snapshot@entry=0x0, seqServerHost=seqServerHost@entry=0x185b2ab "127.0.0.1", 
    seqServerPort=seqServerPort@entry=3803) at pquery.c:842
#11 0x000000000084a5d9 in exec_mpp_query (query_string=query_string@entry=0x185ad9a "select count(*) from n1;", serializedQuerytree=serializedQuerytree@entry=0x0, 
    serializedQuerytreelen=serializedQuerytreelen@entry=0, serializedPlantree=serializedPlantree@entry=0x185adb3 "M\f", serializedPlantreelen=serializedPlantreelen@entry=699, 
    serializedParams=serializedParams@entry=0x0, serializedParamslen=serializedParamslen@entry=0, serializedSliceInfo=serializedSliceInfo@entry=0x185b06e "\003\a", 
    serializedSliceInfolen=serializedSliceInfolen@entry=573, seqServerHost=seqServerHost@entry=0x185b2ab "127.0.0.1", seqServerPort=seqServerPort@entry=3803, localSlice=localSlice@entry=1)
    at postgres.c:1271
#12 0x000000000084f95d in PostgresMain (argc=<optimized out>, argv=argv@entry=0x17632c8, dbname=0x1763230 "bench", username=<optimized out>) at postgres.c:4824
#13 0x00000000007e247a in BackendRun (port=0x176a5d0) at postmaster.c:6699
#14 BackendStartup (port=0x176a5d0) at postmaster.c:6386
#15 ServerLoop () at postmaster.c:2450
#16 0x00000000007e37b9 in PostmasterMain (argc=argc@entry=15, argv=argv@entry=0x173f050) at postmaster.c:1526
#17 0x0000000000487edf in main (argc=15, argv=0x173f050) at main.c:206

strace:
recvmsg(30, 

lsof:
postgres 34339 gpadmin   30u  netlink              0t0      294025 ROUTE


sysrq_trigger:
Mar  6 18:23:59 sdw4 kernel: postgres        S ffff88280b663000     0 34339  31988 0x00000080
Mar  6 18:23:59 sdw4 kernel: ffff8827e414b9e0 0000000000000082 ffff88280bb47300 ffff8827e414bfd8
Mar  6 18:23:59 sdw4 kernel: ffff8827e414bfd8 ffff8827e414bfd8 ffff88280bb47300 0000000000000000
Mar  6 18:23:59 sdw4 kernel: ffff8827e414bb70 ffff88280b663090 ffff88280b663090 ffff88280b663000
Mar  6 18:23:59 sdw4 kernel: Call Trace:
Mar  6 18:23:59 sdw4 kernel: [<ffffffff8163a879>] schedule+0x29/0x70
Mar  6 18:23:59 sdw4 kernel: [<ffffffff81638569>] schedule_timeout+0x209/0x2d0
Mar  6 18:23:59 sdw4 kernel: [<ffffffff8129d98d>] ? context_struct_compute_av+0x34d/0x470
Mar  6 18:23:59 sdw4 kernel: [<ffffffff810c3f02>] ? enqueue_task_fair+0x402/0x6c0
Mar  6 18:23:59 sdw4 kernel: [<ffffffff8151f618>] __skb_recv_datagram+0x4e8/0x5a0
Mar  6 18:23:59 sdw4 kernel: [<ffffffff810c34a7>] ? enqueue_entity+0x237/0x890
Mar  6 18:23:59 sdw4 kernel: [<ffffffff8151f110>] ? datagram_poll+0x100/0x100
Mar  6 18:23:59 sdw4 kernel: [<ffffffff8151f711>] skb_recv_datagram+0x41/0x60
Mar  6 18:23:59 sdw4 kernel: [<ffffffff81557aa2>] netlink_recvmsg+0x62/0x450
Mar  6 18:23:59 sdw4 kernel: [<ffffffff815113df>] sock_recvmsg+0xbf/0x100
Mar  6 18:23:59 sdw4 kernel: [<ffffffff810c1976>] ? dequeue_entity+0x106/0x510
Mar  6 18:23:59 sdw4 kernel: [<ffffffff810c218e>] ? dequeue_task_fair+0x40e/0x620
Mar  6 18:23:59 sdw4 kernel: [<ffffffff810bb675>] ? sched_clock_cpu+0x85/0xc0
Mar  6 18:23:59 sdw4 kernel: [<ffffffff8101360b>] ? __switch_to+0x17b/0x4b0
Mar  6 18:23:59 sdw4 kernel: [<ffffffff8108fddb>] ? recalc_sigpending+0x1b/0x50
Mar  6 18:23:59 sdw4 kernel: [<ffffffff815116ee>] ___sys_recvmsg+0x11e/0x2b0
Mar  6 18:23:59 sdw4 kernel: [<ffffffff81014417>] ? do_signal+0x57/0x6c0
Mar  6 18:23:59 sdw4 kernel: [<ffffffff810929df>] ? ptrace_notify+0x5f/0xc0
Mar  6 18:23:59 sdw4 kernel: [<ffffffff815122a1>] __sys_recvmsg+0x51/0x90
Mar  6 18:23:59 sdw4 kernel: [<ffffffff815122f2>] SyS_recvmsg+0x12/0x20
Mar  6 18:23:59 sdw4 kernel: [<ffffffff816458c9>] system_call_fastpath+0x16/0x1b

**Comment by hlinnaka** [comment]
On 07/03/16 09:26, xiong-gang wrote:
> update the glibc patch and the problem is still there.
>
> @hlinnaka here's some information about the problem, do you have any idea about it?
>
> call stack:
> #0  0x00007f2cadb881fd in recvmsg () from /lib64/libc.so.6
> #1  0x00007f2cadbaaa6d in make_request () from /lib64/libc.so.6
> #2  0x00007f2cadbaaf64 in __check_pf () from /lib64/libc.so.6
> #3  0x00007f2cadb71029 in getaddrinfo () from /lib64/libc.so.6
> #4  0x00000000009fb949 in setupUDPListeningSocket (listenerSocketFd=listenerSocketFd@entry=0x18e839c, listenerPort=listenerPort@entry=0x7fffc3da7070, txFamily=txFamily@entry=0x18e83a0)
>      at ic_udpifc.c:1232
> ...

Googling for "recvmsg getaddrinfo hang" took me to 
https://bugzilla.redhat.com/show_bug.cgi?id=1226534, which looks very 
similar. Discussion there suggests that this is a kernel bug that has 
been fixed in recent kernel versions. At a quick glance, I couldn't 
figure out which versions were affected and where it's fixed, but 
upgrading the kernel probably would help.

- Heikki

**Comment by xiong-gang** [comment]
@hlinnaka 
Thanks.
The environment is redhat 7.2, linux kernel version 3.10.0. 
Few days back, i tried to back port the two patches mentioned in the thread, but the two patches are in 4.0.x branch, not in master branch, which means it's a bug in 4.0.x branch only.

**Comment by hlinnaka** [comment]
On 07/03/16 10:44, xiong-gang wrote:
> @hlinnaka
> Thanks.
> The environment is redhat 7.2, linux kernel version 3.10.0.
> Few days back, i tried to back port the two patches mentioned in the thread, but the two patches are in 4.0.x branch, not in master branch, which means it's a bug in 4.0.x branch only.

I'm not so sure. Following the link from that thread to 
http://comments.gmane.org/gmane.linux.network/363085, it was originally 
reported on Linux 3.17, 3.18 and 3.19, too.

If it's not that kernel bug, then we're back to square one, and we need 
to find what's causing this. Can you write a small self-contained 
self-program to reproduce this?

- Heikki

**Comment by xiong-gang** [comment]
No, i tried some ways but can't reproduce it.
My next step would be back port the patches or update to 4.0.9.
Thanks

**Comment by xiong-gang** [comment]
Updated to kernel 4.4.4, the problem is gone.
The patch applied to linux kernel master in 4.2.
We opened a ticked for Redhat requesting back port them to 3.10.

I think this is done here.

**Comment by ivannovick** [comment]
Can you give us minimim required kernel versions to avoid this patch?  We should to the documentation? @mkiyama

**Comment by xiong-gang** [comment]
It's fixed in kernel V4.2

**Comment by mkiyama** [comment]
@ivannovick FYI - I did not receive an email notification for this issue from github.

**Pull request closed without merging**